IBM/ibm-common-service-operator
# support advance topology by copying default CommonService CR to each watch_namespace

### Context

When we are deployed in advanced topology, there might be multiple copies of ibm common service operator in the same scope. Consider following diagram, The Common Service Operator in `tenant1-cp4ba-namespace` and `tenant1-cp4i-namespace` somehow determine that they are redundant with the REAL relevant common services operator in the `tenant1-services-namespace`.

![multi-namespacesx drawio](https://user-images.githubusercontent.com/29827416/220244150-3f40283b-015b-4958-89d4-b3f76c0c3cee.png)

- The MASTER CommonService CR is ALWAYs installed in the same namespace as that specified for the “operatorNamespace” in the CR, as well as in the same namespace as the REAL operator.
- The CommonService CR in the `tenant1-services-namespace` is the MASTER. The CommonService CRs in `tenant1-cp4ba-namespace` and `tenant1-cp4i-namespace` are leaves – they are not the MASTER.
- On initial configuration, the common services operator that co-resides with the MASTER CommonService CR copies this CR to all other namespaces in the scope.
     - This is done by examining the `WATCH_NAMESPACE` environment variable in the operator and copying the CR to all namespaces being watched (other than the current).
     - If new namespaces are added to the scope, the operator is restarted, will examine its scope and recopy the Common Service CR to all other namespaces in the scope – assuming that it determines that it is the MASTER for the scope.
```
name: common-service
namespace: tenant1-services-namespace
...
operatorNamespace: tenant1-services-namespace
servicesNamespace: tenant1-services-namespace
...
``` 
This describes the processing for the common service operator in the `tenant1-services-namespace`

When any common services operator comes up it looks for an existing CommonService CR in its namespace.  It does NOT create the CR if it already exists.  If the “operatorNamespace” in the existing CR does not match the existing namespace, the operator goes dormant.
```
name: common-service
namespace: tenant1-cp4i-namespace
...
operatorNamespace: tenant1-services-namespace
servicesNamespace: tenant1-services-namespace
...
```
This describes the processing for the common service operators in tenat1-cp4ba-namepsace and tenant-cp4i-namespace.

### How to test
- Deploy `future` CatalogSource for other components' installation

- Deploy following CatalogSource which only contains common service operator
```
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: cs-test-operators
  namespace: openshift-marketplace
spec:
  displayName: IBMCS only Operators
  image: 'quay.io/daniel_fan/common-service-operator-catalog:e52b708d-dirty'
  publisher: IBM
  sourceType: grpc
  updateStrategy:
    registryPoll:
      interval: 45m
```

- Create two namespaces - `cloud-pak` and `common-service`
   - `cloud-pak` is the leaf namespace for Cloud Paks
   - `common-service` is the operator and services namespaces

- Install Restricted Namespace Scope Operator in `common-service` namespace

- [Authorize NSS operator the permission in `cloud-pak` namespace](https://github.com/IBM/ibm-namespace-scope-operator#authorization-and-permissions)
```
scripts/authorize-namespace.sh cloud-pak -to common-service
```

- Create following NamespaceScope CR in `common-service` namespace
```
apiVersion: operator.ibm.com/v1
kind: NamespaceScope
metadata:
  name: namespacescope
  namespace: common-service
spec:
  csvInjector:
    enable: true
  namespaceMembers:
    - common-service
    - cloud-pak
  restartLabels:
    intent: projected
```

- Install common service operator in `common-service` namespace

- Configure CommonService CR with correct `CatalogName` in order to deploy ODLM
```
apiVersion: operator.ibm.com/v3
kind: CommonService
metadata:
  name: common-service
  namespace: common-service
spec:
  catalogName: opencloud-operators
  operatorNamespace: common-service
  servicesNamespace: common-service
  size: starterset
```

### How to verify
- Check the deployment of common-service operator
   - make sure that it contains following label and env variable
```
labels:
    intent: projected
    name: ibm-common-service-operator
.......
env:
    - name: WATCH_NAMESPACE
      valueFrom:
        configMapKeyRef:
          key: namespaces
          name: namespace-scope
          optional: true
```
- Check if there is a CommonService CR called `common-service` in `cloud-pak` namespace. It is copied from Master CommonService CR. The copied CommonService CR has following attributes
    - label `operator.ibm.com/common-services.cloned-from: common-service`
    - `.status.configurable: false`
    - CR's namespace `cloud-pak` is different from `.spec.operatorNamespace: common-service`
    - Or CR's namespace `cloud-pak` is different from `.status.configStatus.operatorPlane.operatorNamespace: common-service`
```
apiVersion: operator.ibm.com/v3
kind: CommonService
metadata:
  labels:
    operator.ibm.com/common-services.cloned-from: common-service
  name: common-service
  namespace: cloud-pak
spec:
  catalogName: opencloud-operators
  license:
    accept: false
  operatorNamespace: common-service
  servicesNamespace: common-service
  size: starterset
status:
    configStatus:
      catalogPlane:
        catalogName: opencloud-operators
        catalogNamespace: openshift-marketplace
      operatorPlane:
        operatorNamespace: common-service
      servicesPlane:
        servicesNamespace: common-service
  configurable: false
  phase: Succeeded
```
- Install a second common-service-operator in `cloud-pak` namespace.
   - It does not reconcile any resources, instead it goes dormant with following logs
```
I0221 04:14:41.456931 1 main.go:132] Identifying Common Service Operator Role in the namespace cloudpak
I0221 04:14:41.499358 1 main.go:245] Common Service Operator goes dormant in the namespace cloudpak
I0221 04:14:41.499388 1 main.go:246] Common Service Operator in the namespace cpfs-operator takes charge of resource managemet
```
